### PR TITLE
note start/end of interactive periods on document

### LINKED ISSUE #5937: LOD in linked plots
When I have two linked plots (via `x_range=plot1.x_range`)  and move them only the plot I'm moving by hand gets LOD mode applied. This effectively makes the whole moving operation slow again. LOD should be applied to all plot being moved.

Version: 0.12.4

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #5937
- [x] tests added / passed

This PR adds support for "linked" LOD by making the document centrally responsible to note the start/end/duration of interactive periods. This simplifies the code at the point of usage and allows for some simple unit tests to be added. 

With this work, LOD start/end events are only triggered on the plot that originated the interactive period. 

Here is an example script that can be used:
```
import numpy as np

from bokeh.plotting import figure, show, output_file
from bokeh.layouts import row

N = 4000
x = np.random.random(size=N) * 100
y = np.random.random(size=N) * 100
radii = np.random.random(size=N) * 1.5
colors = [
    "#%02x%02x%02x" % (int(r), int(g), 150) for r, g in zip(50+2*x, 30+2*y)
]

p0 = figure()

p0.scatter(x, y, radius=radii,
          fill_color=colors, fill_alpha=0.6,
          line_color=None)

x = np.random.random(size=N) * 100
y = np.random.random(size=N) * 100
radii = np.random.random(size=N) * 1.5
colors = [
    "#%02x%02x%02x" % (int(r), int(g), 150) for r, g in zip(50+2*x, 30+2*y)
]

p1 = figure(x_range=p0.x_range, y_range=p0.y_range)

p1.scatter(x, y, radius=radii,
          fill_color=colors, fill_alpha=0.6,
          line_color=None)

output_file("color_scatter.html", title="color_scatter.py example")

show(row(p0, p1))  # open a browser
```

There are further refinements that could be made in the future. In particular the same simple LOD threshold applies for all plots but this is not always ideal. In particular default settings do not trigger LOD on the second plot in the `vector.py` example, because the data size for the `multi_line` is small (92) even though the total number of points that make up those lines is much larger.

### CODE REVIEW

**Comment by bryevdv** [comment]
Still some tweaks to make sure lod events are handled reasonably

**Comment by bryevdv** [comment]
This is seeming to fail due to two JS examples not working? Not sure what's going on there yet.

**Comment by bryevdv** [comment]
Also three JS tests I thought were passing I had accidentally just commented out. Some problem with sinon/spy but I have not been able to to figure it out. @mattpap if nothing jumps out at you we can just close this PR and reject the issue with `wontfix` it's not that important.

**Comment by mattpap** [comment]
@bryevdv, `gulp examples` fails to compile. I see this is in other PRs as well, so it's unrelated. TypeScript 2.6 was released yesterday, so I presume this is the reason. I will fix this in the morning.

**Comment by bryevdv** [comment]
OK I found the issue, the mocks for `trigger_event` were expecting `calledOnce` to be true, but it's not. However, `called` is, and I believe this is correct. In addition to the actual UI event, e.g. pan or wheel, there is also LODStart. So I think it should expect more than one call, which is what the test now reflects.

**Comment by bryevdv** [comment]
This should be pretty uncontroversial I will merge in the morning and cut a dev build unless there is comment.

File: bokehjs/src/coffee/document.coffee [hunk 1]
```diff
@@ -148,6 +152,23 @@ export class Document
...
       @_pop_all_models_freeze()
 
+  interactive_start: (plot) ->
+    if @_interactive_plot == null
+      @_interactive_plot = plot
+      @_interactive_plot.trigger_event(new LODStart({}))
+    @_interactive_timestamp = Date.now()
+
+  interactive_stop: (plot) ->
+    if @_interactive_plot?.id == plot.id
```
Comment (mattpap) [inline bokehjs/src/coffee/document.coffee]: @bryevdv, what's the reason for comparing ids instead of just references (`@_interactive_plot == plot`)?

**Pull request merged**